front/src/applications/operationalStudies/components/MacroEditor/OsrdNgeSync.tsx

```diff
@@ -0,0 +1,918 @@
+    try {
+      const createPromise = this.dispatch(
+        osrdEditoastApi.endpoints.postProjectsByProjectIdStudiesAndStudyIdScenariosScenarioIdMacroNodes.initiate(
+          {
+            projectId: this.scenario.project.id,
+            studyId: this.scenario.study_id,
+            scenarioId: this.scenario.id,
+            macroNodeForm: node,
+          }
+        )
+      );
+      const newNode = await createPromise.unwrap();
+      this.indexNode(newNode, true);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  private async apiUpdateNode(node: MacroNodeResponse) {
+    try {
+      await this.dispatch(
+        osrdEditoastApi.endpoints.putProjectsByProjectIdStudiesAndStudyIdScenariosScenarioIdMacroNodesNodeId.initiate(
+          {
+            projectId: this.scenario.project.id,
+            studyId: this.scenario.study_id,
+            scenarioId: this.scenario.id,
+            nodeId: node.id,
+            macroNodeForm: node,
+          }
+        )
+      );
+      this.indexNode(node, true);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  private async apiDeleteNode(node: MacroNodeResponse) {
+    try {
+      await this.dispatch(
+        osrdEditoastApi.endpoints.deleteProjectsByProjectIdStudiesAndStudyIdScenariosScenarioIdMacroNodesNodeId.initiate(
+          {
+            projectId: this.scenario.project.id,
+            studyId: this.scenario.study_id,
+            scenarioId: this.scenario.id,
+            nodeId: node.id,
+          }
+        )
+      );
+      this.deleteNodeByKey(node.path_item_key);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  /**
+   * Cast a NGE node to a node.
+   */
+  static castNgeNode(
+    node: NetzgrafikDto['nodes'][0],
+    labels: NetzgrafikDto['labels']
+  ): Omit<MacroNodeResponse, 'path_item_key'> {
+    return {
+      id: node.id,
+      trigram: node.betriebspunktName,
+      full_name: node.fullName,
+      connection_time: node.connectionTime,
+      position_x: node.positionX,
+      position_y: node.positionY,
+      labels: node.labelIds
+        .map((id) => {
+          const ngeLabel = labels.find((e) => e.id === id);
+          if (ngeLabel) return ngeLabel.label;
+          return null;
+        })
+        .filter((n) => n !== null) as string[],
+    };
+  }
+
+  /**
+   * Given an path step, returns its opId
+   */
+  static getPathKey(item: TrainScheduleResult['path'][0]): string {
+    if ('trigram' in item)
+      return `trigram:${item.trigram}${item.secondary_code ? `/${item.secondary_code}` : ''}`;
+    if ('operational_point' in item) return `op_id:${item.operational_point}`;
+    if ('uic' in item) return `uic:${item.uic}`;
+    if ('track' in item && 'offset' in item) return `track_offset:${item.track}+${item.offset}`;
+
+    // Error
+    console.error(`Can't find an path ID for ${JSON.stringify(item)}`);
+    throw new Error(`Can't find an path ID for ${JSON.stringify(item)}`);
```

Comment:
Since to TS, we know that we can not have an error.
I think it's better to write things like this, so if the type `PathItemLocation` changes, we will have directly an error TS.
```suggestion
   return `track_offset:${item.track}+${item.offset}`;
```